File: c/insertion_sort.c
#include <stdio.h>

// Insertion sort
int main(){
    int size = 5;

    int arr[5] = {12, 3, 2, 5, 10};

    int tmp;
    int i = 1;
    while(i < size){
        int key = arr[i];
        int j = i;
        while(j > 0 && arr[j - 1] > arr[j]){
            arr[j] = arr[j - 1];
            j = j - 1;
        }
        arr[j] = key;
        i = i + 1;
    }
    for(int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}


/*
 Complexity:

    Time : O(N^2)
    Memory : O(N)

    Worst case - Swaps : O(N^2)

 */

File: python/insertion_sort.py
# Insertion Sort

"""
arr  =  [12, 3, 2, 5, 10]

step 1: [3, 12, 2, 5, 10]

step 2: [3, 2, 12, 5, 10]

step 3: [2, 3, 12, 5, 10]

step 4: [2, 3, 5, 12, 10]

step 5: [2, 3, 5, 10, 12] -> Sorting Done.

"""

# Insertion sort using shifting (efficient than swapping)
def InsertionSort(arr: list[int]) -> list[int]:
    i = 1
    while i < len(arr):
        key = arr[i]
        j = i - 1
        while j > 0 and arr[j - 1] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
        i += 1
    return arr

"""
	Complexity(worst case):

		Time: O(N^2)
		Memory: O(1)

		Swaps: O(N^2)

"""
